### PR TITLE
Implementation of `choose` and `constant` for #131

### PR DESCRIPTION
**Pull request by metasim**
Initial implementation, so feedback appreciated.

### CODE REVIEW

File: gremlin-scala/src/main/scala/gremlin/scala/GremlinScala.scala [hunk 1]
```diff
@@ -364,6 +364,18 @@ case class GremlinScala[End, Labels <: HList](traversal: GraphTraversal[_, End])
     GremlinScala[A, Labels](traversal.union(mappedTraversals: _*))
   }
 
+  def choose[A, B](
+    predicate: GremlinScala[End, HNil] ⇒ GremlinScala[A, _],
+    trueChoice: GremlinScala[End, HNil] ⇒ GremlinScala[B, _],
+    falseChoice: GremlinScala[End, HNil] ⇒ GremlinScala[B, _]): GremlinScala[B, Labels] = {
+    val p: Traversal[_, A] = predicate(start).traversal
```
Comment (mpollmeier) [inline gremlin-scala/src/main/scala/gremlin/scala/GremlinScala.scala]: shouldn't this be `predicate: GremlinScala[End, HNil] ⇒ GremlinScala[Boolean, _]` as this is a predicate? Then you only need one type parameter

File: gremlin-scala/src/test/scala/gremlin/scala/TraversalSpec.scala [hunk 2]
```diff
@@ -394,6 +395,34 @@ class TraversalSpec extends WordSpec with Matchers {
     }
   }
 
+  "choose" should {
+    "work for if/then/else semantics" in new Fixture {
+
+      val traversal: GremlinScala[String, _] = graph
+        .V.hasLabel("person")
+        .choose(
+          _.values("age").is(P.lte(30)),
```
Comment (mpollmeier) [inline gremlin-scala/src/test/scala/gremlin/scala/TraversalSpec.scala]: you could use the (typed) property Key as defined at the bottom and used in other tests here, i.e. `_.value2(Age)`

File: gremlin-scala/src/test/scala/gremlin/scala/TraversalSpec.scala [hunk 3]
```diff
@@ -394,6 +395,34 @@ class TraversalSpec extends WordSpec with Matchers {
...
+          _.in(),
```
Comment (mpollmeier) [inline gremlin-scala/src/test/scala/gremlin/scala/TraversalSpec.scala]: personally I'd use named parameters here, i.e. `trueChoice = _.in()` - just to make it more readable

File: gremlin-scala/src/main/scala/gremlin/scala/GremlinScala.scala [hunk 4]
```diff
@@ -364,6 +364,18 @@ case class GremlinScala[End, Labels <: HList](traversal: GraphTraversal[_, End])
     GremlinScala[A, Labels](traversal.union(mappedTraversals: _*))
   }
 
+  def choose[A, B](
+    predicate: GremlinScala[End, HNil] ⇒ GremlinScala[A, _],
+    trueChoice: GremlinScala[End, HNil] ⇒ GremlinScala[B, _],
```
Comment (mpollmeier) [inline gremlin-scala/src/main/scala/gremlin/scala/GremlinScala.scala]: should we call the choice parameters `onTrue` and `onFalse`? just thinking aloud here, happy to keep as is if you prefer

File: gremlin-scala/src/test/scala/gremlin/scala/TraversalSpec.scala [hunk 5]
```diff
@@ -394,6 +395,34 @@ class TraversalSpec extends WordSpec with Matchers {
...
+
+      traversal.toSet shouldBe Set("marko", "ripple", "lop", "lop")
+    }
+
+    "work with `constant`" in new Fixture {
+      val traversal: GremlinScala[String, _] = graph
+        .V
+        .choose(
+          _.hasLabel("person"),
+          _.values[String]("name"),
```
Comment (mpollmeier) [inline gremlin-scala/src/test/scala/gremlin/scala/TraversalSpec.scala]: same as above: consider named parameters and `.value(Key)` step

**Comment by mpollmeier** [comment]
that's looking good, thanks. added my 2 cents in the comments

File: gremlin-scala/src/main/scala/gremlin/scala/GremlinScala.scala (hunk 1 above)
Comment (metasim) [inline gremlin-scala/src/main/scala/gremlin/scala/GremlinScala.scala]: I thought the same thing too, but when I looked at the TinkerPop docs, it says that the predicate is considered `true` if the traversal is non-empty.

What's your architecture leaning here: preserve the original Gremlin/TinkerPop semantics, or adapt them to make them more type-consistent?

File: gremlin-scala/src/test/scala/gremlin/scala/TraversalSpec.scala (hunk 2 above)
Comment (metasim) [inline gremlin-scala/src/test/scala/gremlin/scala/TraversalSpec.scala]: Will fix.

File: gremlin-scala/src/test/scala/gremlin/scala/TraversalSpec.scala (hunk 3 above)
Comment (metasim) [inline gremlin-scala/src/test/scala/gremlin/scala/TraversalSpec.scala]: Good idea.

File: gremlin-scala/src/test/scala/gremlin/scala/TraversalSpec.scala (hunk 5 above)
Comment (metasim) [inline gremlin-scala/src/test/scala/gremlin/scala/TraversalSpec.scala]: Will fix.

File: gremlin-scala/src/main/scala/gremlin/scala/GremlinScala.scala (hunk 4 above)
Comment (metasim) [inline gremlin-scala/src/main/scala/gremlin/scala/GremlinScala.scala]: I just used what was in the the TinkerPop Java. I like `onTrue`, etc. better too. Shall I change it?

File: gremlin-scala/src/main/scala/gremlin/scala/GremlinScala.scala (hunk 1 above)
Comment (mpollmeier) [inline gremlin-scala/src/main/scala/gremlin/scala/GremlinScala.scala]: I'm typically leaning towards more type level consistency, and this is a place where I would definitely go for Boolean. Making empty traversal equal true doesn't make sense to me, I'd rather require the user to make a clear statement. And if they care about empty traversals, they can use `exists`

File: gremlin-scala/src/main/scala/gremlin/scala/GremlinScala.scala (hunk 4 above)
Comment (mpollmeier) [inline gremlin-scala/src/main/scala/gremlin/scala/GremlinScala.scala]: yup, change it then.

**Comment by metasim** [comment]
Working on this but having a problem making the `predicate` return `GremlinScala[Boolean, _]` due to downstream types. Expressions such as  `_.value(Age).is(P.lte(30))` are of type `GremlinScala[Int, _]`. To make it work I end up with this mess:

```
      val traversal: GremlinScala[String, _] = graph
        .V.hasLabel("person")
        .choose(
          _.value(Age).is(P.lte(30)).map(_ != 0),
          onTrue = _.in(),
          onFalse = _.out()
        )
        .value(Name)
```

**Comment by mpollmeier** [comment]
Maybe it's easier if we make `predicate: End => Boolean`

The `is` step doesn't change the end value to a boolean, it rather acts like `filter` (Gremlin-Scala simply calls gremlin-java, there is no difference between the two on that step):

```
import gremlin.scala._
import org.apache.tinkerpop.gremlin.tinkergraph.structure.TinkerFactory
import org.apache.tinkerpop.gremlin.tinkergraph.structure.TinkerGraph
import org.apache.tinkerpop.gremlin.process.traversal.dsl.graph.__
import org.apache.tinkerpop.gremlin.process.traversal.{Order, P, Scope}
val g = TinkerFactory.createModern.asScala

g.V.hasLabel("person").value[Int]("age").is(P.lte(30)).toList
res3: List[Int] = List(29, 27)
```

**Comment by metasim** [comment]
Yeh, that's looking better:


```scala
      val traversal = graph
        .V.hasLabel("person")
        .choose(
          _.value2(Age) <= 30,
          onTrue = _.in(),
          onFalse = _.out()
        )
        .value(Name)

      traversal.toSet shouldBe Set("marko", "ripple", "lop", "lop")
...
      val traversal = graph
        .V
        .choose(
          _.label() == "person",
          onTrue = _.value(Name),
          onFalse = _.constant("inhuman")
        )

      traversal.toSet() shouldBe Set("marko", "vadas", "inhuman", "josh", "peter")

```

**Comment by mpollmeier** [comment]
Agreed, looks great! Feel free to merge if you're happy with it.

We might have to rethink the use of lambdas going forward, or at least provide both options (lamda and traversal), as RemoteGraph doesn't work with lambdas: https://github.com/apache/incubator-tinkerpop/pull/269 and lambdas are hard to optimise for the database.

**Pull request merged**